nodeJS/express/controllers.md

```diff
@@ -0,0 +1,451 @@
+
+Ultimately, controllers are pretty much just functions with well-defined responsibilities as part of the MVC pattern, that's it.
+
+### Lesson overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- Explore various response methods
+- Implement error handling middleware to catch and process errors gracefully
+- Describe common use cases for middleware, such as validation and authentication
+- Know basic file structure
+
+### Handling responses
+
+When it comes to sending responses from your controllers, you have several methods at your disposal. Let's explore some of the commonly used methods and their use cases.
+
+- [res.send](https://expressjs.com/en/api.html#res.send) - A general-purpose method for sending a response, it is flexible with what data you can send since it will automatically set the `Content-Type` header based on what data you pass, and yes including setting to an `application/json`.
+- [res.json](https://expressjs.com/en/api.html#res.json) - If you want to send a JSON response, `res.json` is the way to go. It automatically sets the `Content-Type` header to `application/json` and sends the data as JSON. We will cover the use-case of this more later.
+- [res.redirect](https://expressjs.com/en/api.html#res.redirect) - When you want to redirect the client to a different URL, this method allows for that capability.
+- [res.render](https://expressjs.com/en/api.html#res.render) - If you're using a template engine that you will setup in a later lesson, `res.render` allows you to render a view template and send the rendered HTML as the response.
+
+There is also a useful method that you can use to set the status code manually.
+
+- [res.status](https://expressjs.com/en/api.html#res.status) - You can also chain other methods through this (e.g. `res.status(200).send(...)` but note that you can't do `res.send(...).status(200)`). This method does not end the request-response cycle. And also `200` is already the default so you wouldn't necessarily need this for `200`, but for consistency, we'll keep using it.controller
+
+<div class="lesson-note" markdown="1">
+
+#### res.send and res.json
+
+If `res.send` automatically sets the `Content-Type` based on the data passed. Why still use `res.json`? It's because `res.json` enforces JSON and will automatically convert non-object to JSON but `res.send` will not. But `res.json` is just a convenient method that also internally calls `res.send` but it handles anything JSON related. `res.send` also calls `res.json` internally but you will only see JSON type of `Content-Type` with booleans and objects.
+
+So for convenience it's more appropriate to use `res.json` instead of `res.send`, and if you are sending JSON then you might as well go for a method that is *named "json"*  :)
```

Comment:
Are we fine on the smilies in lessons? Nothing inherently wrong with it of course, just wondering if it's better to stick with proper punctuating and using tone for lightheartedness/being casual.

Reply:
rephrased it to:

> So for convenience it's more appropriate to use `res.json` instead of `res.send`, and if you're sending JSON, you might as well use a method that's literally *named "json"*! It's like the perfect match!

let me know what you think

Reply:
Perfecto 👌